x86 data cache fix

tbursztyka opened the pull request:
cache management in x86 seemed pretty fragile, if not broken.

Commit:
arch/x86: Fixing dcache enable/disable code

It did not build in x86_64 due to the fact that cr0 is a 64bits
register in such architecture, instead of being a 32bits one originaly
so the place holder has to follow that size. Such place holder must be
initialized to 0 to make sure no upper 32 bits ends up set which would
conclude in a general protection error.

Operand size specifier (l, q ...) is useless as well in this context.

Clearing up the masks by using proper macros.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: CPUs expose an unimplemented feature

Instruction cache related functions (see include/zephyr/arch/cache.h)
are not implemented in x86, thus let's not set CONFIG_CPU_HAS_ICACHE
which may end up in build failure if one sets CONFIG_ICACHE.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/kernel: Add qemu_x86/x86_64 to cache test and enable cache

How cache test could be ran without CONFIG_CACHE_MANAGEMENT?

Adding qemu_x86/x86_64 as cache related API in x86 has proven to be
very brittle against changes recently: no test was at least verifying
it builds properly, not to mention running it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

finikorg commented on arch/x86/core/cache.c:
```diff
@@ -18,32 +18,46 @@
+	uint64_t cr0 = 0;
+#else
 	uint32_t cr0;
+#endif
```
Can we better use here `unsigned long` ?

tbursztyka replied:
indeed, tough initialization is only valid for x86_64. I wonder if we have a stricter type for it though (such as uintptr_t )

finikorg commented on arch/x86/core/cache.c:
```diff
@@ -18,32 +18,46 @@
 #include <stdbool.h>
 
+#define X86_REG_CR0_NW BIT(29)
+#define X86_REG_CR0_CD BIT(30)
```
I would add comments for those flags

tbursztyka replied:
sure

finikorg commented on tests/kernel/cache/testcase.yaml:
```diff
@@ -8,6 +8,7 @@ tests:
       - qemu_xtensa
       - qemu_cortex_a53
       - nsim_em
+      - qemu_x86_64
```
need to add `qemu_x86`

tbursztyka replied:
I did not add qemu_x86 since it did not work (a new commit fixes it)

finikorg reviewed (commented):
Some comments, otherwise looks good

Commit:
boards/x86: Adding clflush CPU capability to qemu_x86

Without this, cache manipulation cannot work in x86.
clflush was introduced with SSE2 extension, but may be implemented
without this extension and it seems that qemu made this choice a
adding sse2 CPU capability does not include clflush. It enabled by
default in 64 bits, so let's take care of this capability only for
the 32bits version of the board.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented on arch/x86/core/cache.c:
```diff
@@ -18,32 +18,46 @@
+			 "mov %0, %%cr0;\n\t"
+			 : "=r" (cr0)
+			 : "i" (~(X86_REG_CR0_NW | X86_REG_CR0_CD)));
 }
```
No this PR is about fixing cache management, not factorizing cr0 access

finikorg commented on boards/x86/qemu_x86/board.cmake:
```diff
@@ -56,6 +56,9 @@ endif()
 if(CONFIG_X86_SSE4A)
   string(JOIN "," QEMU_CPU_FLAGS "${QEMU_CPU_FLAGS}" "sse4a")
 endif()
+if((NOT CONFIG_X86_64) AND CONFIG_CACHE)
```
Should we check for `CACHE_MANAGEMENT` ? I think braces are not needed in `NOT ...`

tbursztyka replied:
CONFIG_CACHE is for driver, yes, not the actual api.

tbursztyka commented:
huh, now that the test enables CACHE_MANAGEMENT to **actually** test the cache api, it fails for some boards which either have their implementation broken or non-existent.

- bcm958402m2_m7
- adp_xc7k_ae350

Looks like this needs patching as well

finikorg commented:
> huh, now that the test enables CACHE_MANAGEMENT to **actually** test the cache api, it fails for some boards which either have their implementation broken or non-existent.
> 
>     * bcm958402m2_m7
> 
>     * adp_xc7k_ae350
> 
> 
> Looks like this needs patching as well

I propose blacklist the broken boards

finikorg commented on tests/kernel/cache/prj.conf:
```diff
@@ -1,2 +1,3 @@
 CONFIG_ZTEST=y
 CONFIG_TEST_USERSPACE=y
+CONFIG_CACHE_MANAGEMENT=y
```
Now we, probably, need to remove `filter: CONFIG_CACHE_MANAGEMENT`

tbursztyka replied:
I don't think so, because the prj.conf sets CONFIG_CACHE_MANAGEMENT, but it might stay deselected if targets do not have CPU_HAS_DCACHE/CPU_HAS_ICACHE (which respectively make DCACHE/ICACHE selected by default, these being the dependency for CACHE_MANAGEMENT).

finikorg commented:
> > I propose blacklist those boards. Or we can remove enabling CACHE_MANAGEMENT for all boards and make it default to x86. Blacklisting is better since it would clearly indicate that those boards do not support cache management, now the test passed OK.
> 
> We could do that yes, I anyway opened 2 issues #65863 and #65870

If we blacklist those boards, we could merge this PR earlier.

Commit:
tests/kernel: Exlude bogus platforms on cache test

Until these get fixed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@finikorg @jhedberg @andyross Please re-review. Looks like it's all fine now. Thanks
